Fix empty collections not showing in carousel in Full Mode.

pjft opened the pull request:
Hi.

This fixes the empty collections missing from the carousel in Full Mode since Kids mode was introduced.

@Zigurana - these are fairly small changes, but since I do not really know the intention behind the changes I'd really appreciate your comments and approval in this case. I tried to mitigate the impact by forcing the new behavior to only take place in Full Mode as that's really only when you can edit the collections.

If we can edit favorites in Kiosk mode, maybe this needs to be changed to support that otherwise if ES starts in Kiosk mode with empty favorites, adding a game to Favorites will not make it show. I don't know/recall what's the intended behavior in Kiosk mode, so let me know.

Thanks!

Edit: I also make the change to the gamePos < 1 thing here.

Edit 2: I was making that change, but the issue then is that showing empty Collections in Kiosk mode is meaningless as they can't be edited there. So leaving that only for the case when Favorites start empty probably needs either a very specific fix (i.e. only allowing it for "Favorites" in that mode), or just accepting this for now.

We'll chat next week as it might be that - as you suggested - this is legacy code and the comparison is no longer needed. :)

Cheers.

joolswills commented:
Is everyone happy with this?

pjft commented:
I think @zigurana was still going to review this to confirm it doesn't change the way kiosk mode is intended to work. Sorry about that.

zigurana commented on es-app/src/SystemData.cpp:
```diff
@@ -365,7 +366,7 @@ SystemData* SystemData::getNext() const
 		if (it == sSystemVector.cend())
 			it = sSystemVector.cbegin();
-	} while ((*it)->getDisplayedGameCount() == 0); 
+	} while ((*it)->getDisplayedGameCount() == 0 && !(UIModeController::getInstance()->isUIModeFull() && (*it)->isCollection()));
```
So, the plain-language version of this code is: while the system aught to be hidden, go to the next / previous system.
In the past, this was a simple criterion: if there is anything to show, it should not be hidden, Now however, its a bit more complicated, (and might even be more complicated in the future).
In any case, I think that this method should have less knowledge on these details, but rather should be able to ask: "should this system be visible now?", So I would propose to create a new method: 
```
bool isVisible()
{
   return (getDisplayedGameCount() > 0 || 
           !(UIModeController::getInstance()->isUIModeFull() && (*it)->isCollection()));
}
```
That same method can then be called by getPrevious(), and during population (SystemView line 40).

zigurana commented on es-app/src/views/SystemView.cpp:
```diff
@@ -37,7 +37,7 @@ void SystemView::populate()
 			getViewElements(theme);
 
-		if((*it)->getDisplayedGameCount() > 0)
+		if((*it)->getDisplayedGameCount() > 0 || (UIModeController::getInstance()->isUIModeFull() && (*it)->isCollection()))
```
see my comment above

zigurana commented on es-app/src/views/gamelist/BasicGameListView.cpp:
```diff
@@ -111,13 +111,13 @@ void BasicGameListView::remove(FileData *game, bool deleteFile)
 		std::vector<FileData*> siblings = parent->getChildrenListToDisplay();
 		auto gameIter = std::find(siblings.cbegin(), siblings.cend(), game);
-		int gamePos = (int)std::distance(siblings.cbegin(), gameIter);
+		unsigned int gamePos = (int)std::distance(siblings.cbegin(), gameIter);
```
I have no real comment here, just that it took me (too) long to grok. Maybe it's because I am used to read Exceptions-first, default behaviour-last? I dont know. but never mind, it's fine as is.

pjft commented:
Thanks for the review and the comments. I will look into these changes in the near future, they are great recommendations.

Cheers!

pjft commented:
Now that I have a working development set up I'll pick this up shortly.

pjft commented:
Sorry about the tremendous notification spam (and thanks Tomaz82 for the heads up!).

I believe I incorporated the changes @zigurana suggested (thank you, as always!), and I also tracked down a few smaller side effects caused by the UI modes on empty collections.

I believe this should be good to go, but would appreciate some sanity checking - especially on @zigurana 's side as this effectively changes his code.

Thanks.

pjft commented on es-app/src/CollectionSystemManager.cpp:
```diff
@@ -208,6 +208,7 @@ void CollectionSystemManager::updateSystemsList()
 	{
 		if ((*sysIt)->isCollection())
 		{
+			(*sysIt)->getIndex()->setUIModeFilters();
```
@zigurana any objections or considerations to its use here - or anywhere in particular? I take it that this just enforces the UI-mode-specific filters, correct?

Best.

zigurana replied:
Yes, that's correct, but now that I look at it, why should you (as in CollectionSystemManager) care? maybe we should add it to ViewController->getGamelistview(), let me think about that for a bit.

pjft commented:
Rebased, moved the UI filters to ViewController per @zigurana 's suggestion.

Honestly, I don't quite remember what was the edge case I was trying to address there back in January, but I wasn't able to find any outstanding issues so I'm happy with this as it is.

Feedback appreciated as usual.

pjft commented:
Gently pinging this, @zigurana , if/when you have the chance.

Hope all is well with you - haven't heard from you in a while.

joolswills commented:
If he's not around and no-one else has any issues I'm happy to merge this. Thanks.

pjft commented:
Hi folks. Any comments or objections on this? 

Cheers.

joolswills commented:
thanks.
